### PR TITLE
feat(optimizer): support covered index selection for index join

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- As `BatchLookupJoin` has already supported the lookup prefix of the order key, we can support covered index selection for lookup join right now.
- When we try to convert a logical join into a batch lookup join, we will enumerate all the indexes of the lookup table and we prefer the index table can provide longer lookup prefix length.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/6589

### CODE REVIEW

**Review by BowenXiao1999 — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -403,17 +403,83 @@ impl LogicalJoin {
...
         logical_join: LogicalJoin,
-    ) -> Option<PlanRef> {
+    ) -> Option<BatchLookupJoin> {
         match logical_join.join_type() {
-            JoinType::RightOuter
-            | JoinType::RightSemi
-            | JoinType::RightAnti
-            | JoinType::FullOuter => return None,
-            _ => {}
+            JoinType::Inner | JoinType::LeftOuter | JoinType::LeftSemi | JoinType::LeftAnti => {}
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:481]: So we need to transform "right join" to  "right join" if we can?

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:481]: if you mean transform "right join" to "left join". yes.

**Pull request merged**